Add delegation feature to PKI realm

albertzaharovits opened the pull request:
Introduces the `allow_delegation` PKI Realm setting and the `X509AuthenticationToken.isDelegated` property and puts them both at work inside `TransportDelegatePkiAction` to implement the exchange of a certificate chain for an ES access token, aka PKI delegation.

If `allow_delegation` is `true` (default `false`) on a PKI realm then it will authenticate `X509AuthenticationToken`s created by the `TransportDelegatePkiAction` which have the `delegated` property set.

It is a follow-up and hinges on https://github.com/elastic/elasticsearch/pull/43932
Relates #34396

elasticmachine commented:
Pinging @elastic/es-security

albertzaharovits commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/pki/PkiAuthDelegationIntegTests.java:
```diff
@@ -0,0 +1,84 @@
+        return false; // enable http
+    }
+
+    public void testDelegatePki() throws Exception {
```
@bizybot I am not claiming that this is an exhaustive test, and I haven't yet raised the PR for the REST action, but this integ test I believe captures the whole flow. I wonder, though, what are your plans for the tests proxy? Couldn't we add more tests to this class and claim full test coverage for the feature? What is the proxy required for?

albertzaharovits commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/pki/PkiRealmSettings.java:
```diff
@@ -37,6 +37,10 @@
         key -> Setting.intSetting(key, DEFAULT_MAX_USERS, Setting.Property.NodeScope));
 
+    public static final Setting.AffixSetting<Boolean> ALLOW_DELEGATION_SETTING = Setting.affixKeySetting(
+            RealmSettings.realmSettingPrefix(TYPE), "allow_delegation",
```
Thanks. I have changed `allow_delegation` to `delegation.enabled`. I hope that's suitable?

albertzaharovits commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/DelegatePkiRequest.java:
```diff
@@ -0,0 +1,92 @@
+    }
+
+    @Override
+    public ActionRequestValidationException validate() {
```
To me, this looks like we would be duplicating checks without a clear reason for it.

We validate the chain (the order, the signatures, the attributes of every cert, and that the last one is trusted) inside `PkiRealm#authenticate` using JDK's internal ` javax.net.ssl.X509TrustManager.checkClientTrusted` . Why do you think the order, specifically, should be prior validated in the request validation? We would be clearly documenting the format of the array, including the order requirement, similar to this https://tools.ietf.org/html/rfc7515#section-4.1.6 , and array's order are preserved in JSONs.

> our logic of extracting the username would not be correct.

Our logic of extracting the principal would not be correct if any validation criterion is violated, not only the order.

albertzaharovits replied:
Wow! You're right! I was not aware un-ordered "chains" are validated by the X509TrustManager.
I have made the change to validate the order, both in the request, and also during the X509AuthenticationToken building. Thanks!

tvernum commented:
I don't follow what's going on here with the feature branch...

The development is being done directly on the main repo, and then we're proposing to merge it to `master`. That doesn't seem right.

I expected to see development on private branches, with PRs to merge into the feature branch on the main repo.
I don't think this should be merged to `master`. It introduces an incomplete feature that we don't want to ship _as-is_ (because there aren't enough protections around the delegation), so we should hold off merging it until it's in a shippable form.

albertzaharovits commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/DelegatePkiRequest.java:
```diff
@@ -0,0 +1,109 @@
+                return false;
+            }
+            prevIssuer = cert.getIssuerX500Principal();
+        }
```
LGTM, thanks for the suggestion!

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/TransportDelegatePkiAction.java:
```diff
@@ -0,0 +1,73 @@
+
+import java.util.Map;
+
+public class TransportDelegatePkiAction extends HandledTransportAction<DelegatePkiRequest, DelegatePkiResponse> {
```
`TransportDelegatePkiAuthenticationAction` sounds best to me. I'll go with that.

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/pki/X509AuthenticationToken.java:
```diff
@@ -44,4 +53,20 @@ public String dn() {
+            }
+            prevIssuer = cert.getIssuerX500Principal();
+        }
+    }
```
Thanks for suggesting `CertParsingUtils` !

albertzaharovits commented:
@tvernum @bizybot I have addressed your reviews, please take another look :)

> I don't follow what's going on here with the feature branch...

There is a story:

The plan was to divide  #43796 into several pieces. In order to do that, and have PRs based on one another (chained), PRs must be branches in the origin repo. Because the first piece #43932 was a bug fix, it was based on master, and this PR was based on that. When the bug fix had been merged, I should've picked Yogesh's branch, merge master, and re-target this PR to that; Instead I pointed this one to master by oversight, instead of creating a new branch. I have retargeted this to `proxy-pki`.

albertzaharovits commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ssl/CertParsingUtils.java:
```diff
@@ -275,4 +276,20 @@ public static X509ExtendedTrustManager trustManager(KeyStore keyStore, String al
+    /**
+     * Checks that the {@code X509Certificate} array is ordered, such that the endpoint's is first and it is followed by any certificate
+     * authorities'. The check validates that the {@code issuer} of every certificate is the {@code subject} of the certificate in the next
+     * array position. No other certificate attributes are checked.
```
I dug into the cert path validation code and RFCs. I think we should **not** assert end entity quality on the first certificate.

`sun.security.provider.certpath.ConstraintsChecker` validates the CA attribute for all certs in the chain, but it does **not** check the first one (Same for `KeyChecker`, validating the key usage attribute). In [RFC 5280](tools.ietf.org/html/rfc5280) end entity certificate is loosely described as `"is a certificate that either does not contain a Basic Constraints extension, or that contains such an extension with the CA flag equal to FALSE"` but in another section it is said explicitly that `"Usually, the last certificate is an end entity certificate, but it can be a CA"` 

There is also the practical reason that a self signed certificate from the truststore (be it CA or not) should be "trusted", ie the client can use that and be authenticated.

tvernum replied:
I think we need to decide what our contract is here.
We are trusting that the proxy (Kibana) performed a valid TLS handshake using that certificate.

Do we also want to duplicate the checks that the proxy ought to have performed? (The ones we can check, since we can't check that the end-client holds the proper key).  Maybe we do, but I don't think that's necessarily the case. We could reasonable decide either way.

If node.js understands & validates an extension that the JVM/JSSE doesn't, should that be an error?

The certificate ordering check is reasonable because would extract an incorrect DN/principal if the certs aren't in the right order, but we need to determine whether we want to take on full X.509 validation, given that we already know we can't validate the handshake.

albertzaharovits replied:
There is a distinction to make here.
On one hand there is the question of whether to assert the **lack** of the BasicConstraint, which confers the CA role and it is described here [rfc5280#section-4.2.1.9](https://tools.ietf.org/html/rfc5280#section-4.2.1.9), and on the other hand is if we validate the key usage extension such as it can be used for TLS authentication.

On the first matter, I read the RFC that there is no validation required that the first cert be an end entity. The section that specifically describes this constraint (BasicConstraint) only, is [rfc5280#section-4.2.1.9](https://tools.ietf.org/html/rfc5280#section-4.2.1.9), and it says clearly `This extension MAY appear as a critical or non-critical extension in end entity certificates.` This is in accordance with the excerpts you pasted, because the negation of the constraint is explained in `If the basic constraints extension is not present in a version 3 certificate, or the extension is present but the cA boolean is not asserted, then the certified public key MUST NOT be used to verify certificate signatures.` In a different section [section-6.1](https://tools.ietf.org/html/rfc5280#section-6.1) on the path validation algorithm it is mentioned `In most cases, the target certificate will be an end entity certificate, but the target   certificate may be a CA certificate as long as the subject public key is to be used for a purpose other than verifying the signature on a public key certificate.` .
For me, this means the relationship goes in the opposite direction, if you are a CA you must have this constraint, and not if you have this constraint you must be a CA.

On the second matter, I believe you are right we should be validating the key usage extension, but I haven't looked into that. I was hoping this was already handled? I will be looking into that.

albertzaharovits replied:
@bizybot in `sun.security.validator.Validator#validate` there is
```
        // omit EE extension check if EE cert is also trust anchor
        if (chain.length > 1) {
            // EndEntityChecker does not need to check unresolved critical
            // extensions when validating with a TYPE_PKIX Validator.
            // A TYPE_PKIX Validator will already have run checks on all
            // certs' extensions, including checks by any PKIXCertPathCheckers
            // included in the PKIXParameters, so the extra checks would be
            // redundant.
            boolean checkUnresolvedCritExts =
                    (type == TYPE_PKIX) ? false : true;
            endEntityChecker.check(chain[0], parameter,
                                   checkUnresolvedCritExts);
        }
```
This is where the key usage constraint is checked on the end entity certificate (if it is not a trust anchor).
So the check you are suggesting we perform is already done as part of the chain validation.
_I tried testing it but my openssl foo is weak._

Do you conclude the test is already handled in the validation we already perform and so we don't need to make any extra checks?

albertzaharovits commented:
@bizybot I'll add a test with a proper chain in a follow-up.
